File: Lua/mazes/maze1/maze.lua
#!./lua5.1

--
-- Default arguments
--
local DEFAULT_WIDTH = 10
local DEFAULT_HEIGHT = 10
local DEFAULT_SEED = os.time()

--
-- Required imported files
-- (see other Lua libs in this branch)
--
dofile "../../bit/bit.lua"		-- provides bitwise operator services
dofile "../../oo/classlib.lua"		-- provides object inheritance services

-- 
-- Constant values needed for calculations
--
local N,S,E,W = 1,2,4,8
local DX = {}; DX[E] = 1; DX[W] = -1; DX[N] = 0; DX[S] = 0;
local DY = {}; DY[E] = 0; DY[W] = 0; DY[N] = -1; DY[S] = 1;
local OPPOSITE = {}
OPPOSITE[W] = E; OPPOSITE[E] = W;
OPPOSITE[N] = S; OPPOSITE[S] = N;

--[[
	Class Maze defines basic behavior to which a maze should conform.
	It provides basic initialization/construction for the maze class,
	and provides a method for drawing ASCII mazes.

	Specific "maze-carving" techniques are implemented in subclasses.
]]
class.Maze()
function Maze:__init(w,h,s)
	
	-- configure instance variables
	self.width = w
	self.height = h
	self.seed = s

	-- seed the PRNG
	math.randomseed(self.seed)

	-- create the maze "grid" itself
	self.grid = {}
	for j = 1, self.height do
		self.grid[j] = {}
		for i = 1, self.width do
			self.grid[j][i] = 0
		end
	end
end

--[[
	Draw the grid, starting in the upper-left hand corner.
]]
function Maze:draw()
	
	-- draw the "top" line
	lines = {}; out = " ";
	for i = 1,  2*self.width-1 do
		out = out .. "_"	
	end
	table.insert(lines,out)

	-- draw the rows of the maze
	for j = 1, self.height do
		out = "|"
		for i = 1, self.width do

			-- draw "south" wall
			if ( bit.band(self.grid[j][i],S) ~= 0 ) then
				out = out .. " "
			else
				out = out .. "_"
			end

			-- draw "east" wall
			if ( bit.band(self.grid[j][i],E) ~= 0 ) then
				if ( bit.band( bit.bor( self.grid[j][i], self.grid[j][i+1] ), S ) ~= 0 ) then
					out = out .. " "
				else
					out = out .. "_"
				end
			else
				out = out .. "|"
			end
		end
		table.insert(lines,out)
	end

	-- output maze metadata
	out = "[arg]" .. " " .. self.width .. " " .. self.height .. " " .. self.seed
	table.insert(lines,out)

	print(table.concat(lines,"\r\n"))
end

-- Accessor: width
function Maze:width()
        return self.width
end
-- Accessor: height
function Maze:height()
	return self.height
end
-- Accessor: seed
function Maze:seed()
	return self.seed
end

--[[
	Class BackTracker implements a simple recursive back-tracking algorithm
	to draw ASCII mazes. The algorithm works as a "depth-first" search
	of the "tree" or "graph" representing the maze.

	A possible optimization might be to implement a "breadth-first" search.
]]
class.BackTracker(Maze)
function BackTracker:__init(w,h,s)

	-- invoke the super constructor
	self.Maze:__init(w,h,s)

	-- carve the maze grid
	self.carve_passage_from(0,0)

end

function BackTracker:carve_passage_from(x,y)
	print("carving")
end

--[[
	Parse command line options
]]
maze = BackTracker(10,10,DEFAULT_SEED)
maze:draw()

File: PHP/mazes/maze1/maze.php
#!/usr/bin/php
<?php

$DEFAULT_WIDTH = 10;
$DEFAULT_HEIGHT = 10;
$DEFAULT_SEED = make_seed();

// random seed generator for PHP
function make_seed()
{
	list($usec,$sec) = explode(' ',microtime());
	return (float)$sec + ((float)$usec * 100000);
}

// ==================================================================
// Class Maze defines basic behavior to which a maze should conform.
// It provides basic initialization/construction for the maze class, 
// and provides a method for drawing ASCII mazes.
//
// Specific "maze-carving" techniques are implemented in subclasses.
// ==================================================================
class Maze
{
	// static variables indicating directions
	public static $N = 1;
	public static $S = 2;
	public static $E = 4;	
	public static $W = 8;

	// wrap DX, DY, OPPOSITE as static functions, since otherwise hard to get, e.g., self::$E reference
	public static function DX() { return array(self::$E => +1, self::$W => -1, self::$N => 0, self::$S => 0); }
	public static function DY() { return array(self::$E => 0, self::$W => 0, self::$N => -1, self::$S => +1); }
	public static function OPPOSITE() { return array(self::$E => self::$W, self::$W => self::$E, self::$N => self::$S, self::$S => self::$N ); }

	// instance variables for maze
	public $width;
	public $height;
	public $seed;

	// data structure for maze itself
	protected $grid;

	// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	// Construct a new 2D maze with the given width and height.
	//
	// Default seed value will give "random" behavior.
	// User-supplied seed value will give deterministic behavior.
	// +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
	function __construct($w=NULL, $h=NULL, $s=NULL) {
		// import the global variables
		global $DEFAULT_WIDTH, $DEFAULT_HEIGHT, $DEFAULT_SEED;

		// configure the instance variables
		$this->width = is_null($w) ? $DEFAULT_WIDTH : $w;
		$this->height = is_null($h) ? $DEFAULT_HEIGHT : $h;
		$this->seed = is_null($s) ? $DEFAULT_SEED : $s;

		// seed the PRNG
		srand($this->seed);
 
		// initialize two-dimensional grid representing maze 
		$this->grid = array();
		for ( $j=0; $j < $this->height; ++$j ) {
			$temp = array();
			for ( $i=0; $i < $this->width; ++$i ) {
				array_push($temp,0);
			}
			array_push($this->grid,$temp);
		}
	}

	// +++++++++++++++++++++++++++++++++++++++++++++++++++++++
	// Draw the grid, starting in the upper-left hand corner.
	// +++++++++++++++++++++++++++++++++++++++++++++++++++++++
	function draw() {
		// draw the top row
		$buffer = array();
		$out = " ";
		for ( $i=0; $i < (2 * $this->width - 1); ++$i ) {
			$out .= "_";
		}
		array_push($buffer,$out);

		// draw each row of the maze
		for ( $j=0; $j < $this->height; ++$j ) {
			$out = "|";
			for ( $i=0; $i < $this->width; ++$i ) {
				// render the bottom 
				$out .= (($this->grid[$j][$i] & self::$S) != 0) ? " " : "_";

				// render the side
				if ( ($this->grid[$j][$i] & self::$E) != 0 ) {
					$out .= ( ( $this->grid[$j][$i] | $this->grid[$j][$i+1] ) & self::$S ) != 0 ? " " : "_";
				} else {
					$out .= "|";
				}
			}
			array_push($buffer,$out);
		}

		// draw maze metadata + the last row
		$temp = array($this->width,$this->height,$this->seed);
		array_push($buffer,join($temp," "));
		array_push($buffer,"");

		// output the buffer
		echo join($buffer,"\r\n");
	}
}

class BackTracker extends Maze
{
	function __construct($w=NULL, $h=NULL, $s=NULL) {
		//
		// Invoke super-constructor
		//
		parent::__construct($w,$h,$s);

		//
		// Carve the grid
		//
		self::carve_passage_from(0,0);
	}

	function carve_passage_from($x,$y) {
		// randomize the directions
		$directions = array(self::$N, self::$S, self::$E, self::$W);
		shuffle($directions);
		
		$DX = self::DX();
		$DY = self::DY();
		$OPPOSITE = self::OPPOSITE();

		for ( $i=0; $i < sizeof($directions); ++$i ) {
			$direction = $directions[$i];
			$dx = $x + $DX[$direction];
			$dy = $y + $DY[$direction];
			if ( $dy >= 0 && $dy < $this->height && $dx >= 0 && $dx < $this->width && $this->grid[$dy][$dx] == 0 ) {
				$this->grid[$y][$x] |= $direction;
				$this->grid[$dy][$dx] |= $OPPOSITE[$direction];
				self::carve_passage_from($dx,$dy);
			}
		}
	}
}

$maze = new BackTracker();
$maze->draw();
?>
